keel-web/src/main/kotlin/com/netflix/spinnaker/keel/rest/AuthorizationSupport.kt

```diff
@@ -18,59 +18,73 @@
+    READ, WRITE;
+    override fun toString() = name.toLowerCase()
   }
 
-  fun userCanModifyApplication(application: String): Boolean = try {
-    val deliveryConfig = repository.getDeliveryConfigForApplication(application)
-    userCanModifyDeliveryConfig(deliveryConfig.serviceAccount, application, deliveryConfig.name)
-  } catch (e: NoSuchElementException) {
-    // If there are no delivery configs for that application return true so this is handled correctly in the controller.
-    true
+  enum class Entity {
+    APPLICATION, DELIVERY_CONFIG, RESOURCE;
+    override fun toString() = name.toLowerCase()
   }
 
-  fun userCanModifyDeliveryConfig(serviceAccount: String, application: String, manifestName: String): Boolean {
-    val auth = SecurityContextHolder.getContext().authentication
-    return userCanAccessServiceAccount(auth, serviceAccount, "Delivery config for application $application: $manifestName")
-  }
+  data class Permission(val action: Action, val entity: Entity)
 
-  fun userCanAccessServiceAccount(auth: Authentication, serviceAccount: String, specOrName: Any): Boolean {
-    val hasPermission = permissionEvaluator.hasPermission(auth, serviceAccount, "SERVICE_ACCOUNT", "ignored-svcAcct-auth")
-    log.debug(
-      "[AUTH] {} is trying to access service account {}. They{} have permission. {}",
-      auth.principal,
-      serviceAccount,
-      if (hasPermission) "" else " DO NOT",
-      specOrName
-    )
-    return hasPermission
+  fun userCan(action: Action, entity: Entity, identifier: String): Boolean {
+    return try {
+      val auth = SecurityContextHolder.getContext().authentication
+      val serviceAccount = when (entity) {
+        Entity.RESOURCE -> repository.getResource(identifier).serviceAccount
+        Entity.APPLICATION -> repository.getDeliveryConfigForApplication(identifier).serviceAccount
+        Entity.DELIVERY_CONFIG -> repository.getDeliveryConfig(identifier).serviceAccount
+      }
+      val application = when (entity) {
+        Entity.RESOURCE -> repository.getResource(identifier).application
+        Entity.APPLICATION -> identifier
+        Entity.DELIVERY_CONFIG -> repository.getDeliveryConfig(identifier).application
+      }
+      val hasServiceAccountAccess = permissionEvaluator.hasPermission(auth, serviceAccount, "SERVICE_ACCOUNT", "ignored")
+        .also {
+          log.debug("[ACCESS {}] User {}: access to service account {}.",
+            if (it) "ALLOWED" else "DENIED", auth.principal, serviceAccount)
+        }
+      val hasApplicationAccess = permissionEvaluator.hasPermission(serviceAccount, application, "APPLICATION", action.name)
+        .also {
+          log.debug("[ACCESS {}] Service account {}: permission to {} {} in application {}.",
+            if (it) "ALLOWED" else "DENIED", serviceAccount, action.name, "${entity.name} $identifier", application)
+        }
+      hasServiceAccountAccess && hasApplicationAccess
+    } catch (e: NoSuchEntityException) {
```

Comment:
This is where `NoSuchEntityException` as the ancestor of all "not found"-type repository exceptions comes into play.